handler/src/main/java/io/netty/handler/ssl/Java9SslUtils.java

```diff
@@ -0,0 +1,186 @@
+
+        if (!wrapped.getUseClientMode()) {
+            installSelector(engine, applicationNegotiator);
+        }
+    }
+
+    static String getApplicationProtocol(SSLEngine sslEngine) {
+        try {
+            return (String) GET_APPLICATION_PROTOCOL.invoke(sslEngine);
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void setApplicationProtocols(SSLParameters parameters, List<String> supportedProtocols) {
+        String[] protocolArray = supportedProtocols.toArray(EmptyArrays.EMPTY_STRINGS);
+        try {
+            SET_APPLICATION_PROTOCOL.invoke(parameters, new Object[]{ protocolArray });
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static void installSelector(final JdkSslEngine sslEngine,
+                                        final JdkApplicationProtocolNegotiator applicationProtocolNegotiator) {
+        checkNotNull(applicationProtocolNegotiator, "applicationProtocolNegotiator");
+
+        Object biFunction = AccessController.doPrivileged(new PrivilegedAction<Object>() {
+            @Override
+            public Object run() {
+                return Proxy.newProxyInstance(Java9SslUtils.class.getClassLoader(), new Class[]{ BI_FUNCTION_CLASS },
+                        new AlpnSelector(sslEngine, applicationProtocolNegotiator));
+            }
+        });
+
+        try {
+            SET_HANDSHAKE_APPLICATION_PROTOCOL_SELECTOR.invoke(sslEngine.getWrappedEngine(), biFunction);
+        } catch (Exception ex) {
+            throw new IllegalStateException(ex);
+        }
+    }
+
+    private static final class AlpnSelector implements InvocationHandler {
+        private final SSLEngine jdkSslEngine;
+        private final JdkApplicationProtocolNegotiator applicationNegotiator;
+        private final Set<String> protocols;
+
+        AlpnSelector(JdkSslEngine jdkSslEngine, JdkApplicationProtocolNegotiator applicationNegotiator) {
+            this.jdkSslEngine = jdkSslEngine;
+            this.applicationNegotiator = applicationNegotiator;
+            protocols = Collections.unmodifiableSet(new LinkedHashSet<String>(applicationNegotiator.protocols()));
+        }
+
+        @SuppressWarnings("unchecked")
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) throws Throwable {
+            if (!"apply".equals(method.getName())) {
+                throw new UnsupportedOperationException("Invoked unknown method " + method.getName());
+            }
+
+            JdkApplicationProtocolNegotiator.ProtocolSelector selector = applicationNegotiator.protocolSelectorFactory()
+                    .newSelector(jdkSslEngine, protocols);
+
+            try {
+                String selected = selector.select((List<String>) args[1]);
+                return selected == null ? StringUtil.EMPTY_STRING : selected;
+            } catch (Exception cause) {
```

Comment:
We actually throw an exception if we want it to fail for example:

https://github.com/netty/netty/blob/4.1/handler/src/main/java/io/netty/handler/ssl/JdkBaseApplicationProtocolNegotiator.java#L205